Feature/explore why land is rendered as sea

memsom opened the pull request:
This adds in code to fix the background colour issues for the various land masses that should have been rendered as cutouts in blocks of water, and also attempts to rotate text that is drawn upside down to the correct orientation where possible.

memsom commented:
I plan to merge this when the PR build succeeds.

memsom commented on source/Mapping/VexTile.MbTiles.Mvt.TestApp/MainWindow.axaml.cs:
```diff
@@ -12,7 +13,7 @@ public MainWindow()
 
         var connectionString = new SQLiteConnectionString("zurich.mbtiles", SQLiteOpenFlags.ReadOnly, false);
-        var source = new MvtVectorTileSource(connectionString);
+        var source = new MvtVectorTileSource(connectionString, whitelist: ["water"]);
```
This needs to be reverted

memsom commented on source/AliFlux/VexTile.Renderers.Mvt.AliFlux/SkiaCanvas.cs:
```diff
@@ -456,13 +474,9 @@ private double GetPathLength(List<Point> path)
 
-    private double GetAbsoluteDiff2Angles(double x, double y, double c = Math.PI) =>
-        c - Math.Abs((Math.Abs(x - y) % 2 * c) - c);
+    private double GetAbsoluteDiff2Angles(double x, double y, double c = Math.PI) => c - Math.Abs((Math.Abs(x - y) % 2 * c) - c);
```
use expression body

memsom commented on source/AliFlux/VexTile.Renderers.Mvt.AliFlux/SkiaCanvas.cs:
```diff
@@ -516,7 +530,23 @@ public void DrawTextOnPath(List<Point> geometry, Brush style)
+        // is the path | ----> | or | <---- | ?
+
+        var ltr = IsLeftToRight(geometry);
+        var ttb = IsTopToBottom(geometry);
```
remove `ttb` as it is not being used and will add a small amount of latency

memsom reviewed (commented):
Reviewing my changes

pauldendulk commented:
You could also ask copilot to review. It already found two (minor) bugs for me. It also helps you to look at your own code from someone else's perspective.

memsom commented:
> You could also ask copilot to review. It already found two (minor) bugs for me. It also helps you to look at your own code from someone else's perspective.

Interesting - I asked copilot and it was happy with the changes. Do you do this in an IDE or through the Github site? I tried the latter. I'm not sure how useful what it says is, but it did summarise the changes nicely:

-----
## Overall Review
* Unit Tests: Added a new unit test to ensure functionality for rendering with specific layers.
* Debugging: Introduced new debugging capabilities with DEBUG_EXTENT and USE_DEBUG_BOX.
* Code Quality: Improved the readability and maintainability by adding utility methods and refactoring existing methods.
* New Features: Added functionality to handle layer whitelists and override background colors during rendering.
* Warnings Suppression: Suppressed specific NuGet warnings in the test project.

The changes appear to improve the functionality, debugging capabilities, and code quality of the project. The added unit test ensures that new features are tested. Overall, the changes look good and seem well-implemented.

pauldendulk commented:
I pressed the copilot icon on the top right of the pull request on github.com. There are some misses, sometimes nothing useful, but it is an extra check that helps a bit.

memsom commented on source/Mapping/VexTile.MbTiles.Mvt.TestApp/MainWindow.axaml.cs:
```diff
@@ -12,7 +13,7 @@ public MainWindow()
 
         var connectionString = new SQLiteConnectionString("zurich.mbtiles", SQLiteOpenFlags.ReadOnly, false);
-        var source = new MvtVectorTileSource(connectionString);
+        var source = new MvtVectorTileSource(connectionString, whitelist: ["water"]);
```
This is a syntax conflict.

memsom commented on source/AliFlux/VexTile.Renderers.Mvt.AliFlux/SkiaCanvas.cs:
```diff
@@ -53,8 +53,13 @@ public void StartDrawing(double width, double height)
     }
 
-    public void DrawBackground(Brush style) =>
+    SKColor _backgroundColor = SKColors.White;
```
this is a little confused, but this has also been refactored to avoid setting things other than water to the background colour for now.

pauldendulk commented on source/AliFlux/VexTile.Renderers.Mvt.AliFlux/VectorStyle.cs:
```diff
@@ -203,9 +200,8 @@ private static object SimplifyJson(JToken token)
     public Brush[] GetStyleByType(string type, double zoom, double scale = 1)
     {
-        List<Brush> results = new();
+        List<Brush> results = [];
 
```
I also git this feedback as well. Copilot is simply wrong here. It is valid and I prefer the [] in this case.

pauldendulk commented on source/AliFlux/VexTile.Renderers.Mvt.AliFlux/VectorStyle.cs:
```diff
@@ -219,24 +215,21 @@ public Brush[] GetStyleByType(string type, double zoom, double scale = 1)
         }
 
-        return results.ToArray();
+        return [.. results];
```
Here copilot is also wrong. In Mapsui I decided to to not use [..  ] because to me ToArray is more clear.  I changed the setting in the .editorconfig here https://github.com/Mapsui/Mapsui/pull/2951/files. Sorry, for this not important comment, but happen to run into it this morning 😄
